### PR TITLE
Migrate from FxCop to .NET Analyzers

### PR DESCRIPTION
**Pull request by jnyrup**
## IMPORTANT 

* [ ] If the PR touches the public API, the changes have been approved in a separate issue with the "api-approved" label.
* [x] The code complies with the [Coding Guidelines for C#](https://www.csharpcodingguidelines.com/).
* [x] The changes are covered by unit tests which follow the Arrange-Act-Assert syntax and the naming conventions such as is used [in these tests](../tree/develop/Tests/FluentAssertions.Equivalency.Specs/MemberMatchingSpecs.cs#L51-L430).
* [ ] If the PR adds a feature or fixes a bug, please update [the release notes](../tree/develop/docs/_pages/releases.md) with a functional description that explains what the change means to consumers of this library, which are published on the [website](https://fluentassertions.com/releases).
* [ ] If the PR changes the public API the changes needs to be included by running [AcceptApiChanges.ps1](../tree/develop/AcceptApiChanges.ps1) or [AcceptApiChanges.sh](../tree/develop/AcceptApiChanges.sh).
* [ ] If the PR affects [the documentation](../tree/develop/docs/_pages), please include your changes in this pull request so the documentation will appear on the [website](https://www.fluentassertions.com/introduction).
    * [ ] Please also run `./build.sh --target spellcheck` or `.\build.ps1 --target spellcheck` before pushing and check the good outcome

Migrate away from the deprecated [Microsoft.CodeAnalysis.FxCopAnalyzers](https://www.nuget.org/packages/Microsoft.CodeAnalysis.FxCopAnalyzers) over to [Microsoft.CodeAnalysis.NetAnalyzers](https://www.nuget.org/packages/Microsoft.CodeAnalysis.NetAnalyzers/) using these [guidelines](https://learn.microsoft.com/en-gb/visualstudio/code-quality/migrate-from-fxcop-analyzers-to-net-analyzers?view=vs-2019).

According to the guide:
> Starting in Visual Studio 2019 16.8 and .NET 5.0, these analyzers are [included with the .NET SDK](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/overview). If you don't want to move to the .NET 5+ SDK or if you prefer a NuGet package-based model, the analyzers are also available in the Microsoft.CodeAnalysis.NetAnalyzers [NuGet package](https://www.nuget.org/packages/Microsoft.CodeAnalysis.NetAnalyzers). You might prefer a package-based model for on-demand version updates.

### CODE REVIEW

File: .editorconfig [hunk 1]
```diff
@@ -134,11 +134,31 @@ dotnet_diagnostic.CA1062.severity = suggestion
 dotnet_code_quality.CA1062.exclude_extension_method_this_parameter = true
 dotnet_code_quality.exclude_extension_method_this_parameter = true
 dotnet_code_quality.null_check_validation_methods = ThrowIfArgumentIsNull
+# CA1031: Do not catch general exception types
```
Comment (jnyrup) [inline .editorconfig:137]: Except CA1851 all these are migrated from Rules.ruleset.

File: Directory.Build.props [hunk 2]
```diff
@@ -4,4 +4,39 @@
     <CheckEolTargetFramework>false</CheckEolTargetFramework>
     <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
   </PropertyGroup>
+
+  <!-- To reduce build times, we only enable analyzers for the newest TFM -->
```
Comment (jnyrup) [inline Directory.Build.props:8]: Moved inclusion and configuration of analysis here to share them between all projects

File: Tests/FluentAssertions.Equivalency.Specs/DataSpecs.cs [hunk 3]
```diff
@@ -591,12 +591,16 @@ public enum ChangeType
...
+    public static IEnumerable<object[]> AllChangeTypesWithAcceptChangesValues
+    {
+        get
+        {
+            return
+                from changeType in Enum.GetValues(typeof(ChangeType)).Cast<ChangeType>()
+                from acceptChanges in new[] { true, false }
+                select new object[] { changeType, acceptChanges };
+        }
+    }
```
Comment (jnyrup) [inline Tests/FluentAssertions.Equivalency.Specs/DataSpecs.cs:603]: A simpler way to make a Cartesian product and fixes analyzers about `changeType` and `acceptChanges` parameters in `changeType => true` and `acceptChanges => true` being unused.

File: Src/FluentAssertions/Execution/LateBoundTestFramework.cs [hunk 4]
```diff
@@ -16,8 +16,10 @@ public void Throw(string message)
 
         if (exceptionType is null)
         {
+#pragma warning disable CA2201 // Do not raise reserved exception types
```
Comment (jnyrup) [inline Src/FluentAssertions/Execution/LateBoundTestFramework.cs:19]: I kind of agree with this analyzer, but not sure which exception type we should throw instead 🤔

Comment (dennisdoomen) [reply Src/FluentAssertions/Execution/LateBoundTestFramework.cs:19]: `NotSupportedException`?

**Review by dennisdoomen — APPROVED** [review]
Nice work

**Pull request merged**